### PR TITLE
ENH: Create `For instructors` page

### PR DESCRIPTION
**Pull request by jsheunis**
This PR aims to address https://github.com/psychoinformatics-de/rdm-course/issues/40:
- describes how to set up [The Littlest Jupyter Hub](https://tljh.jupyter.org/en/latest/index.html) on AWS, mostly with info from @mslw's notes [here](https://github.com/psychoinformatics-de/rdm-course/issues/19).
- adds the content under `Extras > For Instructors`
- links to this content from the `Setup` page

### CODE REVIEW

File: _extras/for_instructors.md [hunk 1]
```diff
@@ -0,0 +1,312 @@
+---
+layout: page
+title: For Instructors
+---
+
+To save time during a workshop, organisors and instructors can consider setting
```
Comment (mslw) [inline _extras/for_instructors.md]:
```suggestion
To save time during a workshop, organizers and instructors can consider setting
```

File: _extras/for_instructors.md [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+4. [Some notes and further options](#4-some-notes-and-further-options)
+
+
+### Before you begin
+
+1. Sign up for an [AWS account](https://aws.amazon.com/) and verify it.
+2. Consider the costs involved in running a cloud-compute server. AWS provides resources
+for checking [EC2 on-demand pricing](https://aws.amazon.com/ec2/pricing/on-demand/) and
+also for [calculating cost estimates](https://calculator.aws/) for your usage.
+*TODO: add ballpark figures for costs of running a workshop for X users for Y days.*
```
Comment (mslw) [inline _extras/for_instructors.md]:
```suggestion
```
We can add this ballpark estimate to the "discussion" box below

File: _extras/for_instructors.md [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+3. If you want to serve your workshop content via a custom domain (e.g.
+[datalad-hub.inm7.de](http://datalad-hub.inm7.de/)), ensure that you have the required
+level of access that will allow you to associate the EC2-provided IP address with your
+chosen domain address.
+
+
+> ## AWS account usage can incur costs
+>
+> While Amazon provides [Free Tier](https://aws.amazon.com/free/) access to its services, it can still potentially result in costs if usage exceeds [Free Tier Limits](https://docs.aws.amazon.com/awsaccountbilling/latest/aboutv2/free-tier-limits.html). Be sure to take note of these limits, or set up [automatic tracking alerts](https://docs.aws.amazon.com/awsaccountbilling/latest/aboutv2/tracking-free-tier-usage.html) to be notified before incurring unnecessary costs.
+{: .discussion}
```
Comment (mslw) [inline _extras/for_instructors.md:40]:
```suggestion

In 2022, our costs for a two half-day workshop were about 15 € (we used a small AWS instance for setup, and manually started & stopped a large instance for the workshop sessions only, i.e. about 8 hours).
{: .discussion}
```
See previous comment

File: _extras/for_instructors.md [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+According to AWS, "free tier eligible customers can get up to 30 GB of EBS General Purpose (SSD) or Magnetic storage".
+8. `Advanced details`: expand this section, and at the very end in the text-field `User data`
+paste the following code (provided in [The Littlest JupyterHub documentation](https://tljh.jupyter.org/en/latest/topic/customizing-installer.html#topic-customizing-installer)):
+
+    ~~~
+    #!/bin/bash
+    curl -L https://tljh.jupyter.org/bootstrap.py \
+    | sudo python3 - \
+        --admin <admin-username>
+    ~~~
```
Comment (mslw) [inline _extras/for_instructors.md]:
```suggestion
   ~~~
   #!/bin/bash
   curl -L https://tljh.jupyter.org/bootstrap.py \
   | sudo python3 - \
       --admin <admin-username>
   ~~~
```
I am not sure what exactly is the mechanism at play here, but with the current indent level, after rendering to html, the whole code block has a leading space in each line. When I copy-pasted the snippet from the html page, the startup script failed, leaving the running server without Jupyter Hub.

To make things weirder, I could ssh into the server and execute the user script without making any changes...

But anyway, with one space less, the code block has no leading spaces, and copy-pasting into AWS worked for me.

File: _extras/for_instructors.md [hunk 5]
```diff
@@ -0,0 +1,312 @@
...
+    {: .discussion}
+
+8. Finally, select `Launch instance`. Give it about 10 minutes to complete the installation.
+
+
+## 2. Create a constant IP address and domain
+
+Once launched and running, the instance with JupyterHub environment will be accessible via a
+public IP address. You can verify this by navigating to `Instances` in the left-hand sidebar of the EC2 console, then selecting your newly launched instance, and clicking on `open address` next to the IP address in the information block. This should open a browser tab/window with a login interface for JupyterHub.
+
```
Comment (mslw) [inline _extras/for_instructors.md:88]:
```suggestion
At this moment, we only have http available (not https).

```
AWS gives you https link by default, so you may need to remove the "s" and/or fight your browser redirection.

File: _extras/for_instructors.md [hunk 6]
```diff
@@ -0,0 +1,312 @@
...
+Technically, this IP address can suffice as a point of entry for users, but there are certain caveats:
+- we ideally want the IP address to stay constant even if we stop and restart our instance
+- we want the information transfer to and from the server to be securely encrypted via HTTPS,
+- and we want to assign a domain to our hub that makes it easily identifiable (e.g. `datalad-hub.inm7.de`)
+
+The Littlest JupyterHub also requires a custom constant domain if HTTPS acces is to be configures. For these reasons, we need an "Elastic IP":
```
Comment (mslw) [inline _extras/for_instructors.md]:
```suggestion
- we ideally want the IP address to stay constant even if we stop and restart our instance,
- we want to assign a domain to our hub that makes it easily identifiable (e.g. `datalad-hub.inm7.de`),
- we want the information transfer to and from the server to be securely encrypted via HTTPS,
- HTTPS access in TLJH can only be configured when we have a domain (i.e. not when using a plain IP address).

For these reasons, we need an "Elastic IP":
```

I'd rephrase as above

File: _extras/for_instructors.md [hunk 7]
```diff
@@ -0,0 +1,312 @@
...
+This ensures that you have the required tools for the rest of the configuration procedure
+as well as those required for the workshop.
+
+Run the following lines one by one, typing `y` for "Yes" whenever you are prompted to continue:
+~~~
+apt update
+apt upgrade
+apt install zsh
+sudo -E conda install git
+sudo -E apt-get install tig
```
Comment (mslw) [inline _extras/for_instructors.md]:
```suggestion
sudo apt update
sudo apt upgrade
sudo apt install zsh tig tree
```

- all `apt` commands above require sudo
- `sudo -E` is neeed for pip & conda, not apt
- I suggest skipping `conda install git` - I needed it when I first worked with Ubuntu 18.04 LTS (Bionic Beaver). Since we recommend 22.04 LTS (Jammy Jellyfish), we get `git 2.34.1` which is plenty new (I initially cared about having `init.defaultBranch` setting introduced in git 2.28; but I also stopped using it because GIN then treats git-annex branch by default...)
- let's add `tree` at this point, it's not installed by default

File: _extras/for_instructors.md [hunk 8]
```diff
@@ -0,0 +1,312 @@
...
+~~~
+
+For the workshop itself, we want to install datalad and its dependencies.
+Run the following lines one by one:
+~~~
+sudo -E apt-get install git-annex
+sudo -E -H pip3 install datalad
```
Comment (mslw) [inline _extras/for_instructors.md:221]: - What does `-H` do?
- With the default setup, `pip` is equivalent to `pip3`, but nothing wrong with being explicit

File: _extras/for_instructors.md [hunk 9]
```diff
@@ -0,0 +1,312 @@
...
+apt upgrade
+apt install zsh
+sudo -E conda install git
+sudo -E apt-get install tig
+~~~
+
+For the workshop itself, we want to install datalad and its dependencies.
+Run the following lines one by one:
+~~~
+sudo -E apt-get install git-annex
```
Comment (mslw) [inline _extras/for_instructors.md]:
```suggestion
sudo -E apt install git-annex
```

Comment (mslw) [reply _extras/for_instructors.md]: **By the way**, at the moment, Jammy Jellyfish provides git-annex 8.20210223 which is old-ish, but I can't say whether we rely on anything added since. So let's leave it for now as-is.

I used to install git-annex & datalad through conda (conda-forge?) to get new stuff. If we want to, we could add neurodebian at the beginning, and get a much newer release from apt. My way of adding neurodebian repo and declaring signing keys is:
```
sudo -i
gpg --no-default-keyring --keyring /usr/share/keyrings/neurodebian-archive-keyring.gpg --keyserver hkps://keyserver.ubuntu.com --recv-keys 0xA5D32F012649A5A9
printf "deb [signed-by=/usr/share/keyrings/neurodebian-archive-keyring.gpg] https://neurodebian.inm7.de/debian jammy main\n" | tee /etc/apt/sources.list.d/inm7.sources.list
exit
```
for some reason, `sudo gpg ...` doesn't behave the same way, hence `sudo -i` before, and `exit` after.

I verified this in AWS, but we can leave it out for now.

File: _extras/for_instructors.md [hunk 10]
```diff
@@ -0,0 +1,312 @@
...
+   git clone https://github.com/sindresorhus/pure.git "$HOME/.zsh/pure"
+   ~~~
+2. Create a `.zshrc` file in your HOME directory and copy the content from [this source](https://gist.github.com/mslw/926f1191e61ef2d705fadab66d19b8ba) into the file:
+   ~~~
+   touch "$HOME/.zshrc"
+   nano "$HOME/.zshrc"
+   ~~~
+   This opens the `.zshrc` file for editing in the shell using nano text editor.
+   Now paste the content from the source file in this editor, then hit CTRL+O and Enter (saves edits)
+   and then hit CTRL+X and Enter (exits editor).
```
Comment (mslw) [inline _extras/for_instructors.md]:
```suggestion
2. Create a `.zshrc` file in your HOME directory with a basic configuration taken, e.g., from [this source](https://gist.github.com/mslw/926f1191e61ef2d705fadab66d19b8ba):
   ~~~
   wget https://gist.githubusercontent.com/mslw/926f1191e61ef2d705fadab66d19b8ba/raw/.zshrc
   ~~~
```
Maybe use wget? The link from above will point to the latest revision.

File: _extras/for_instructors.md [hunk 11]
```diff
@@ -0,0 +1,312 @@
...
+   in order for the changes to take effect.
+
+
+### 3.6. Create a global `gitignore`
+
+When we work with `git` repositories and add content or code, we often want certain files
+or directories not to form part of the `git` history. We can achieve this by telling `git`
+to ignore certain files or directories, via the `.gitignore` configuration. Since development
+in JupyterHub often happens via Jupyter Notebooks, which auto-create and -update it's own metadata
+files in the form of `.ipynb_checkpoints` files, we will add those to the global `gitignore` configuration:
```
Comment (mslw) [inline _extras/for_instructors.md]: The real reason is that these files are created even when working with hub's text editor, not just notebooks.
I suggest people use that editor over nano. But then the checkpoints folder shows up when doing datalad status for the first time. While it can be used as an opportunit to introduce gitignore, I prefer to force this setting silently - the files are, after all, an artifact of Jupyter from our POV.

```suggestion
to ignore certain files or directories, via the `.gitignore` configuration. JupyterHub stores its own
metadata files under `.ipynb_checkpoints` and will create this directory whenever the hub's text
editor is used, or a notebook is opened. To prevent these directories for getting into user's way
when working with DataLad, we will create a global (user-level) `gitignore` configuration following
[this method](https://docs.github.com/en/get-started/getting-started-with-git/ignoring-files#configuring-ignored-files-for-all-repositories-on-your-computer):
```

File: _extras/for_instructors.md [hunk 12]
```diff
@@ -0,0 +1,312 @@
...
+
+~~~
+sudo touch /home/.gitignore_global
+sudo echo ".ipynb_checkpoints" > /home/.gitignore_global
+git config --global core.excludesfile /home/.gitignore_global
```
Comment (mslw) [inline _extras/for_instructors.md]: Did it work for you this way? If so, cool! My idea was to define it on user-level (see link from suggestion above), and make default for users through `/etc/skel`. I verified with bash & zsh that setting the config value as `"~/.gitignore_global"` (with quotes) preserves the tilde in `.gitconfig`, so when copied to `etc/skel`, the file works for other users too.

```suggestion
echo ".ipynb_checkpoints" > ~/.gitignore_global
git config --global core.excludesfile "~/.gitignore_global"
```

File: _extras/for_instructors.md [hunk 13]
```diff
@@ -0,0 +1,312 @@
...
+~~~
+
+### 3.7. Set up default user settings and data
+
+What we have done up until now is to set up the base environment that a user will encounter
+when they log into the JupyterHub. Some tools and packages have been installed for all users
+(as a result of using the `sudo -E` command) such as `git`, `datalad`, and others. However,
+some settings will not yet be the default for any newly created user. This includes the default
```
Comment (mslw) [inline _extras/for_instructors.md]: Before, I suggested dropping the conda install for git, so I'm removing git from examples here. Also rewording a bit, might be a little clearer IMO:

```suggestion
when they log into the JupyterHub. Some tools and packages, such as `datalad`,
 have been installed for all users (as a result of using `pip` with `sudo -E`). However,
some settings were done only for the admin account, and  will not yet be the default for 
other, newly created users. This includes the default
```

File: _extras/for_instructors.md [hunk 14]
```diff
@@ -0,0 +1,312 @@
...
+are available to individual users by copying the required content to the `/etc/skel` directory, the
+contents of which will be placed in the `$HOME` directories of newly created users.
+
+Run the following code to copy the `zsh` configuration files, the `.gitignore_global` file, and the `jupyter_notebook_config.py` file:
+~~~
+sudo mkdir -p /etc/skel/.zsh
+sudo cp -r "$HOME/.zsh/pure" /etc/skel/.zsh/
+sudo cp "$HOME/.zshrc" /etc/skel/
+ 
+sudo cp "$HOME/.gitconfig" /etc/skel/
```
Comment (mslw) [inline _extras/for_instructors.md]: If using my suggestion from 2 comments above (`.gitignore_global` on user level), this would need to become:
```suggestion
sudo cp "$HOME/.gitconfig" /etc/skel/
sudo cp "$HOME/.gitignore_global" /etc/skel/ 
```

File: _extras/for_instructors.md [hunk 15]
```diff
@@ -0,0 +1,312 @@
...
+Run the following code to copy the `zsh` configuration files, the `.gitignore_global` file, and the `jupyter_notebook_config.py` file:
+~~~
+sudo mkdir -p /etc/skel/.zsh
+sudo cp -r "$HOME/.zsh/pure" /etc/skel/.zsh/
+sudo cp "$HOME/.zshrc" /etc/skel/
+ 
+sudo cp "$HOME/.gitconfig" /etc/skel/ 
+
+sudo mkdir -p /etc/skel/.jupyter
+sudo cp "$HOME/.jupyter/jupyter_notebook_config.py" /etc/skel/.jupyter/
```
Comment (mslw) [inline _extras/for_instructors.md:293]: FTR, I finally figured it out. This file can also be placed in
```
/opt/tljh/user/etc/jupyter/jupyter_notebook_config.py
```
to have global effect, making home dir config unnecessary (for regular users and admin). Doesn't matter too much, though.

File: _extras/for_instructors.md [hunk 16]
```diff
@@ -0,0 +1,312 @@
...
+
+Different authentication options are possible (e.g. admin can also authenticate users individually after they sign up, or authentication can happen via Google, GitHub or AWS). Refer to the documentation for further options.
+
+
+## 4. Some notes and further options
+
+### EC2 instance notes
+
+- The letter `a` in type name (eg. `t3a.medium`) means that they are running on AMD, and are slightly cheaper than their Intel counterparts (without the `a`).
+- There are different instance types. We're interested in T3 and M5, both of which are multipurpose. T3 instances are burstable (meaning they accumulate credits when idling and use them when active, so don't offer full performance all the time, which suits us well). M5 are fixed-performance multipurpose (you get the full capacity all the time).
```
Comment (mslw) [inline _extras/for_instructors.md]: The names are cryptic, but this page at least describes them nicely.
```suggestion
- There are different [instance types](https://aws.amazon.com/ec2/instance-types/). We're interested in T3 and M5, both of which are multipurpose. T3 instances are burstable (meaning they accumulate credits when idling and use them when active, so don't offer full performance all the time, which suits us well). M5 are fixed-performance multipurpose (you get the full capacity all the time).
```

File: _extras/for_instructors.md [hunk 17]
```diff
@@ -0,0 +1,312 @@
...
+- For the workshop sessions, consider upscaling to `M5a.4xlarge` (for the live sessions only, meaning downscaling back to `t3a.medium` overnight in a 2-day format). The "quadruple extra large" offers 16 vCPU and 64 GB RAM. Judging by looking at htop, this is serious overprovisioning even for ca 16 people, but at $1 an hour it seems a reasonable price for comfort.
+- To scale (up or down), go to EC2 management console, select your instance, right click (or click on Actions), and stop it. Then again right-click (or Actions), instance settings, change instance type. Then bring it back up.
+- Elastic IP incurs cost when not associated with a running instance, but it is marginal.
+- After the workshop, you will want to pull everything down, including storage (if left sitting, storage may generate a couple dollars per month).
+
+#### TLJH options
+- Increase cull timeout with sudo tljh-config set services.cull.timeout <time in seconds> (you may wish to do this only before the workshop)
```
Comment (mslw) [inline _extras/for_instructors.md]: Maybe worth adding it somewhere earlier?

My thinking was that when the instructor takes a while to explain something, it wouldn't be cool if users got a popup "your server was stopped, click to reload".

File: _extras/for_instructors.md [hunk 18]
```diff
@@ -0,0 +1,312 @@
...
+- We went with `t3a.medium` from the get-go: 2 vCPU and 4GB RAM are very comfortable for all setup steps, and it should be fine for a few concurrent users if you want to leave the hub running between workshop days - the price around $1 a day seems fine for that.
+- For the workshop sessions, consider upscaling to `M5a.4xlarge` (for the live sessions only, meaning downscaling back to `t3a.medium` overnight in a 2-day format). The "quadruple extra large" offers 16 vCPU and 64 GB RAM. Judging by looking at htop, this is serious overprovisioning even for ca 16 people, but at $1 an hour it seems a reasonable price for comfort.
+- To scale (up or down), go to EC2 management console, select your instance, right click (or click on Actions), and stop it. Then again right-click (or Actions), instance settings, change instance type. Then bring it back up.
+- Elastic IP incurs cost when not associated with a running instance, but it is marginal.
+- After the workshop, you will want to pull everything down, including storage (if left sitting, storage may generate a couple dollars per month).
+
```
Comment (mslw) [inline _extras/for_instructors.md:319]: How about a little troubleshooting section? The idea to look at system log comes from TLJH docs, the things to look for I found myself

```suggestion

### Troubleshooting

- When checking if your hub was initialized successfully, apart from trying to access the server's ip address, you may want to view the AWS's system log.
- In EC2 Management Console, Instances page, right click your instance, and select "Monitor and troubleshoot", "Get system log".
- If system setup was finished, there will be a lot of text, most likely ending with something like:
   ~~~
   2022/12/01 12:40:30Z: OsProductName: Ubuntu
   2022/12/01 12:40:30Z: OsVersion: 22.04
   ~~~
- If the JupyterHub bootsrap script succeeded, within the last 30 lines you will find:
   ~~~
   [  210.143720] cloud-init[1233]: Waiting for JupyterHub to come up (1/20 tries)
   [  210.147437] cloud-init[1233]: Done!
   ~~~
- If the "user data" was not given, or pasted incorrectly (e.g. without admin name), within the last 30 lines you will probably see:
   ~~~
   cloud-init[1246]: 2022-12-01 12:40:29,790 - util.py[WARNING]: Running module scripts-user (<module 'cloudinit.config.cc_scripts_user' from '/usr/lib/python3/dist-packages/cloudinit/config/cc_scripts_user.py'>) failed
   ~~~
- In case of problems, you can log in through `ssh` and try running the TLJH bootstrap script manually, or terminate the instance and start over.
```

**Review by mslw — CHANGES_REQUESTED** [review]
Thanks a lot, that's a great work!

I was planning to do such a write-up for a long time, but never actually did, so it's great that you did it.

I left quite a few comments - basically anything is pretty minor, but some small changes to commands would improve copy-pasting and eliminate some errors (e.g.  things like adding `sudo` to `apt`). Some of them go back to the deficiencies in my notes on which you relied, so I also tried to explain a bit more why I made some suggestions, or why I no longer think why some of my older choices are no longer optimal.

File: _extras/for_instructors.md (hunk 10 above)
Comment (jsheunis) [reply _extras/for_instructors.md]: great suggestion

File: _extras/for_instructors.md (hunk 8 above)
Comment (jsheunis) [reply _extras/for_instructors.md:221]: https://www.sudo.ws/docs/man/1.9.12/sudo.man/#H

When installing with pip alongside the conda base environment, I got a warning (that I didn't log somewhere) that suggested using the -H flag. Everything still installed fins without it though.

File: _extras/for_instructors.md (hunk 11 above)
Comment (jsheunis) [reply _extras/for_instructors.md]: thanks for clarifying!

File: _extras/for_instructors.md (hunk 12 above)
Comment (jsheunis) [reply _extras/for_instructors.md]: Ah I missed the quotes initially and that's why it didn't work. Then I changed to `/home/.gitignore_global`. I haven't actually opened JH's text editor yet, so let me test it quickly and will comment here with the outcome.

Comment (mslw) [reply _extras/for_instructors.md]: I guess it may be sufficient for that file to be readable by normal user, but worth checking.

I also finally found git's [gitignore](https://git-scm.com/docs/gitignore) docs:
> Patterns which a user wants Git to ignore in all situations (e.g., backup or temporary files generated by the user’s editor of choice) generally go into a file specified by core.excludesFile in the user’s ~/.gitconfig. Its default value is $XDG_CONFIG_HOME/git/ignore. If $XDG_CONFIG_HOME is either not set or empty, $HOME/.config/git/ignore is used instead.

So maybe we could get away without setting that config option if we put the ignore file in the default location... but I'm probably optimizing too much now ;)

Comment (jsheunis) [reply _extras/for_instructors.md]: wait, there's something like too much optimization?? I need to learn about that 🙃

Let's use your original suggestion to keep it simple. Then users can also find it easily if they want to amend it.

File: _extras/for_instructors.md (hunk 17 above)
Comment (jsheunis) [reply _extras/for_instructors.md]: Do you mean adding this sentence earlier in the document? I see no problem with that. Can you add a code suggestion to put it wherever makes sense for you?

Comment (mslw) [reply _extras/for_instructors.md]: I pushed a commit which moves it higher up to put it after the `tljh-config` commands but before all installation and user space configuration.

**Review by mslw — APPROVED** [review]
Thanks again for putting this page together, very helpful!
